### PR TITLE
gnrc_sixlowpan: Various hardening fixes

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
Title says everything

### Testing procedure
Fragmentation and header compression (e.g. as outlined in Release Specs 4 and 6) should still work.

### CODE REVIEW

**Comment by miri64** [comment]
> Fragmentation and header compression (e.g. as outlined in Release Specs 4 and 6) should still work.

See https://github.com/RIOT-OS/RIOT/actions/runs/3345458318

File: sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c [hunk 1]
```diff
@@ -872,13 +878,12 @@ void gnrc_sixlowpan_iphc_recv(gnrc_pktsnip_t *sixlo, void *rbuf_ptr,
         payload_len = (sixlo->size + uncomp_hdr_len -
                        payload_offset - sizeof(ipv6_hdr_t));
     }
-    if ((rbuf == NULL) &&
-        /* (rbuf == NULL) => forwarding is not affected by this */
-        (gnrc_pktbuf_realloc_data(ipv6, uncomp_hdr_len + payload_len) != 0)) {
+    if (gnrc_pktbuf_realloc_data(ipv6, uncomp_hdr_len + payload_len) != 0) {
```
Comment (benpicco) [inline sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c:886]: No longer conditional to `rbuf` being NULL?

File: sys/net/link_layer/ieee802154/ieee802154.c [hunk 2]
```diff
@@ -115,27 +115,39 @@ size_t ieee802154_set_frame_hdr(uint8_t *buf, const uint8_t *src, size_t src_len
...
+        return len;
+    } else if (tmp != IEEE802154_FCF_TYPE_BEACON) {
+        /* Beacon contains no dst address */
+        tmp = (mhr[1] & IEEE802154_FCF_DST_ADDR_MASK);
+        if (tmp == IEEE802154_FCF_DST_ADDR_SHORT) {
+            len += 4;   /* 2 byte dst PAN + 2 byte dst short address */
+            has_dst = 1;
+        }
+        else if (tmp == IEEE802154_FCF_DST_ADDR_LONG) {
+            len += 10;  /* 2 byte dst PAN + 2 byte dst long address */
```
Comment (benpicco) [inline sys/net/link_layer/ieee802154/ieee802154.c:133]:
```suggestion
            len += 10;  /* 2 byte dst PAN + 8 byte dst long address */
```

File: sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c (hunk 1 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c:886]: Yes, as discussed offline.

Comment (miri64) [reply sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c:886]: But was wrong, as the test showed. Fixes will show the difference.

File: sys/net/link_layer/ieee802154/ieee802154.c (hunk 2 above)
Comment (miri64) [reply sys/net/link_layer/ieee802154/ieee802154.c]: Done.

File: sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c [hunk 3]
```diff
@@ -737,6 +737,12 @@ void gnrc_sixlowpan_iphc_recv(gnrc_pktsnip_t *sixlo, void *rbuf_ptr,
     if (rbuf != NULL) {
         ipv6 = rbuf->pkt;
         assert(ipv6 != NULL);
+        if ((ipv6->size < sizeof(ipv6_hdr_t)) &&
```
Comment (benpicco) [inline sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c:740]: Shouldn't this be

```suggestion
        if ((ipv6->size < sizeof(ipv6_hdr_t)) ||
```

Comment (benpicco) [reply sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c:740]: ah no this tries to embiggen the buffer if it's too small to fit an IPv6 header, right?

File: sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c [hunk 4]
```diff
@@ -754,8 +760,9 @@ void gnrc_sixlowpan_iphc_recv(gnrc_pktsnip_t *sixlo, void *rbuf_ptr,
     iface = gnrc_netif_hdr_get_netif(netif->data);
     payload_offset = _iphc_ipv6_decode(iphc_hdr, netif->data, iface,
                                        ipv6->data);
-    if (payload_offset == 0) {
-        /* unable to parse IPHC header */
+    if ((payload_offset == 0) || (payload_offset > sixlo->size)) {
```
Comment (benpicco) [inline sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c:763]: Not
```suggestion
    if (payload_offset != sixlo->size) {
```
?

File: sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c (hunk 3 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c:740]: Yepp!

File: sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c (hunk 4 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c:763]: No, payload_offset may be smaller than `sixlo->size`. It is, after all, the offset of the payload within `sixlo->data`.

**Review by benpicco — APPROVED** [review]

**Comment by miri64** [comment]
Would be great if someone(TM) could fix this randomly failing test...

**Comment by miri64** [comment]
Mhhh fails pretty reliably on `ci-prod` :-/... but I can not reproduce this locally.

**Comment by miri64** [comment]
Ok, release tests also fail for 6.6 (UDP without payload). Since this touches IPHC and NHC, I wager, this is a regression :-/. Will look into it.

**Comment by miri64** [comment]
1bbac1d5070e8f6347cd9390b337de7671716534 was at fault. This was only introduced after the previous discussion in reaction to the failing tests here. However, other fixes to the "[gnrc_sixlowpan_iphc: fix buffer overflow in gnrc_sixlowpan_iphc_recv()](https://github.com/RIOT-OS/RIOT/pull/18817/commits/73615161c01fcfbbc7216cf502cabb12c1598ee4)" commit, made that fix

1. unnecessary
2. introducing the error in `tests/gcoap_fileserver` and the release specs 6.6 test

As such, I just removed it.

**Comment by miri64** [comment]
> 2\. introducing the error in `tests/gcoap_fileserver`

(I was able to reproduce it after all btw)

**Comment by miri64** [comment]
Release tests are run here: https://github.com/RIOT-OS/RIOT/actions/runs/3346308008/jobs/5548696437

**Comment by miri64** [comment]
> Release tests are run here: https://github.com/RIOT-OS/RIOT/actions/runs/3346308008/jobs/5548696437

Succeeded.

**Comment by miri64** [comment]
Backport provided in #18820

**Pull request merged**